### PR TITLE
feat(Container) : Loading native pytorch weights

### PR DESCRIPTION
**Pull request by mobley-trent**
# PR Description

Adds functionality to load model weights from native Pytorch and Flax formats.

## Related Issue

Closes #

## Checklist

- [ ] Did you add a function?
- [ ] Did you add the tests?
- [ ] Did you run your tests and are your tests passing?
- [ ] Did pre-commit not fail on any check?
- [ ] Did you follow the steps we provided?

### CODE REVIEW

**Comment by mobley-trent** [comment]
Tensorflow supports h5 natively, so I didn't add any feature related to TF. However, I came across issues related to `nvcc-cuda` while trying to implement the relevant functionality for loading Paddle weights @vedpatwardhan

File: ivy/data_classes/container/base.py [hunk 1]
```diff
@@ -1218,6 +1222,66 @@ def cont_from_disk_as_json(json_filepath, ivyh=None):
...
+            results in the global ivy.
+
+        Returns
+        -------
+            Container loaded from disk
+        """
+        from flax.training import checkpoints
+
+        ckpt = checkpoints.restore_checkpoint(
+            ckpt_dir="/workspaces/ivy/load-save-modules/jax_weights", target=None
```
Comment (vedpatwardhan) [inline ivy/data_classes/container/base.py:1244]: Let's replace the hardcoded absolute path here and use the `ckpt_filepath` instead

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
> Tensorflow supports h5 natively, so I didn't add any feature related to TF. However, I came across issues related to `nvcc-cuda` while trying to implement the relevant functionality for loading Paddle weights @vedpatwardhan

Yep sounds good, thanks @mobley-trent! Could you please provide more details about the issue with `nvcc-cuda` when using `paddle` about the setup, etc.? Are you using our gpu docker image for testing (`unifyai/ivy:latest-gpu`)? 
It would also be great to add a couple of tests for these methods similar to [these](https://github.com/unifyai/ivy/blob/91a0b62de5881dcf5e88949262780c252b030e28/ivy_tests/test_ivy/test_misc/test_container.py#L3045). Thanks for creating the PR 😄

**Review by mobley-trent — COMMENTED** [review]
This is the error trace @vedpatwarthan

```
RuntimeError: (PreconditionNotMet) Cannot load cudnn shared library. Cannot invoke method cudnnGetVersion.
  [Hint: cudnn_dso_handle should not be null.] (at /paddle/paddle/phi/backends/dynload/cudnn.cc:64)
```

I'm using the GPU image.

**Comment by vedpatwardhan** [comment]
> I tried running the example from the paddle docs on the `save` function.
> 
> ```python
> import paddle
> emb = paddle.nn.Embedding(10, 10)
> layer_state_dict = emb.state_dict()
> 
> # save state_dict of emb
> paddle.save(layer_state_dict, CKPT_PATH + "emb.pdparams")
> 
> scheduler = paddle.optimizer.lr.NoamDecay(
>     d_model=0.01, warmup_steps=100, verbose=True)
> adam = paddle.optimizer.Adam(
>     learning_rate=scheduler,
>     parameters=emb.parameters())
> opt_state_dict = adam.state_dict()
> 
> # save state_dict of optimizer
> paddle.save(opt_state_dict, CKPT_PATH + "adam.pdopt")
> # save weight of emb
> paddle.save(emb.weight, CKPT_PATH + "emb.weight.pdtensor")
> ```
> 
> This is the error trace @vedpatwardhan
> 
> ```
> RuntimeError: (PreconditionNotMet) Cannot load cudnn shared library. Cannot invoke method cudnnGetVersion.
>   [Hint: cudnn_dso_handle should not be null.] (at /paddle/paddle/phi/backends/dynload/cudnn.cc:64)
> ```
> 
> I'm using the GPU image.

I think we can ignore the error with `paddle` for now, feel free to add the tests and re-request a review once you're done with the changes. Thanks @mobley-trent 😄

**Comment by mobley-trent** [comment]
On the second run of the notebook containing the flax code, the code fails due to some issues with `asyncio`. I was getting this error when trying to load the weights with `flax.checkpoints`:
```
ValueError: loop argument must agree with lock
```

I don't know whether this is an issue with my dev environment, but ultimately I decided to drop the flax method. I think implementing the functionality for loading torch weights was the main objective here. What do you think @vedpatwardhan ?

**Review by vedpatwardhan — COMMENTED** [review]
> On the second run of the notebook containing the flax code, the code fails due to some issues with `asyncio`. I was getting this error when trying to load the weights with `flax.checkpoints`:
> 
> ```
> ValueError: loop argument must agree with lock
> ```
> 
> I don't know whether this is an issue with my dev environment, but ultimately I decided to drop the flax method. I think implementing the functionality for loading torch weights was the main objective here. What do you think @vedpatwardhan ?

Hey @mobley-trent, I'm not sure about the notebook you're referring to, is it one of the notebooks on the `demos` repo? Also, ideally we should be loading the weights synchronously, is the asyncio error an issue with `flax`? Thanks 😄

**Comment by mobley-trent** [comment]
> > On the second run of the notebook containing the flax code, the code fails due to some issues with `asyncio`. I was getting this error when trying to load the weights with `flax.checkpoints`:
> > ```
> > ValueError: loop argument must agree with lock
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I don't know whether this is an issue with my dev environment, but ultimately I decided to drop the flax method. I think implementing the functionality for loading torch weights was the main objective here. What do you think @vedpatwardhan ?
> 
> Hey @mobley-trent, I'm not sure about the notebook you're referring to, is it one of the notebooks on the `demos` repo? Also, ideally we should be loading the weights synchronously, is the asyncio error an issue with `flax`? Thanks 😄

Yes it is an issue with flax. As for the notebook, I am referring to a notebook that I set up myself to test the load feature for flax.

**Review by vedpatwardhan — COMMENTED** [review]
> Yes it is an underlying issue with flax. As for the notebook, I am referring to a notebook that I set up myself to test the load feature for flax.

Is this issue already identified on the `flax` repo? If not, we should create an issue on their repo. Also, are you trying to use an example from their documentation in your notebook, just to confirm that we are using the right code for testing? Thanks @mobley-trent 😄

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Pull request closed without merging**